src/operator-run.c

```diff
@@ -2034,23 +2041,226 @@ void xnn_compute_f16_qd8_convert(
+    } while (0)
+
+#define KAI_UNUSED(x) (void)(x)
+#define KAI_MIN(a, b) (((a) < (b)) ? (a) : (b))
+#define KAI_MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+// The following functions should not be here because this file is not
+// target specific.
+size_t kai_get_lhs_offset_lhs_quant_pack_qai8dxp_f32(
+  size_t m_idx,
+  size_t lhs_stride)
+{
+  return m_idx * lhs_stride;
+}
+
+size_t kai_get_lhs_packed_offset_lhs_quant_pack_qai8dxp_f32(
+  size_t m_idx,
+  size_t k,
+  size_t mr,
+  size_t kr,
+  size_t sr)
+{
+  KAI_ASSERT(k % kr == 0);
+  KAI_UNUSED(kr);
+
+  const size_t kai_num_bytes_per_multiplier = sizeof(float);
+  const size_t kai_num_bytes_per_offset = sizeof(int32_t);
+  const size_t dst_y = (m_idx / mr);
+
+  const size_t dst_stride = mr * (k * sizeof(int8_t) + kai_num_bytes_per_multiplier + kai_num_bytes_per_offset);
+
+  // It points always at the beginning of the row
+  return dst_y * dst_stride;
+}
+
+void kai_run_lhs_quant_pack_qai8dxp_f32(
+  size_t m,
+  size_t k,
+  size_t mr,
+  size_t kr,
+  size_t sr,
+  size_t m_idx_start,
+  const float* restrict lhs,
+  size_t lhs_stride,
+  void* restrict lhs_packed)
+{
+  KAI_ASSERT(k % kr == 0);
+  KAI_ASSERT((kr % sr) == 0);
+
+  const size_t kai_num_bytes_per_multiplier = sizeof(float);
+  const size_t kai_num_bytes_per_offset = sizeof(int32_t);
+
+  if (m == 0) {
+      return;
+  }
+
+  const size_t num_rows = m;
+  const size_t num_cols = k;
+
+  const float* src_ptr = lhs;
+
+  const size_t dst_stride = mr * (k * sizeof(int8_t) + kai_num_bytes_per_multiplier + kai_num_bytes_per_offset);
+  const size_t k_block_len = kr / sr;
+
+  for (size_t row_idx = 0; row_idx < num_rows; ++row_idx) {
+    float32x4_t vmax0 = vdupq_n_f32(-FLT_MAX);
+    float32x4_t vmin0 = vdupq_n_f32(FLT_MAX);
+
+    // Find min/max for each channel
+    size_t k_idx = 0;
+    for (; k_idx <= (num_cols - 8); k_idx += 8) {
+      const float32x4_t src0_0 = vld1q_f32(src_ptr + 0 + k_idx);
+      const float32x4_t src0_1 = vld1q_f32(src_ptr + 4 + k_idx);
+
+      // Calculate the max
+      vmax0 = vmaxq_f32(src0_0, vmax0);
+      vmax0 = vmaxq_f32(vmax0, src0_1);
+
+      // Calculate the min
+      vmin0 = vminq_f32(src0_0, vmin0);
+      vmin0 = vminq_f32(vmin0, src0_1);
+    }
+
+    for (; k_idx < num_cols; ++k_idx) {
+      const float src0_0 = *(src_ptr + k_idx);
+
+      // Calculate the max
+      vmax0 = vsetq_lane_f32(KAI_MAX(src0_0, vgetq_lane_f32(vmax0, 0)), vmax0, 0);
+      // Calculate the min
+      vmin0 = vsetq_lane_f32(KAI_MIN(src0_0, vgetq_lane_f32(vmin0, 0)), vmin0, 0);
+    }
+
+    // Get the max/min
+    const float max0 = vmaxvq_f32(vmax0);
+    const float min0 = vminvq_f32(vmin0);
+
+    // Maximum/minimum int8 values
+    const float qmin = (float)INT8_MIN;
+    const float qmax = (float)INT8_MAX;
+
+    const float rmin0 = KAI_MIN(0.0f, min0);
+    const float rmax0 = KAI_MAX(0.0f, max0);
+
+    const float scale0 = rmin0 == rmax0 ? 1.f : (qmax - qmin) / (rmax0 - rmin0);
```

Comment:
div is slow... consider reciprocal.  Unroll loop to 4 to allow this portion of the code to be simd?